<a href="https://colab.research.google.com/github/fitriindra/MLOps-Zoomcamp-2023/blob/main/MLOps_Zoomcamp_HW_1_Intro.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#MLOps Zoompcamp Homework #1 - Intro

##Q1. Downloading the data
We'll use the same NYC taxi dataset, but instead of "Green Taxi Trip Records", we'll use "Yellow Taxi Trip Records".

Download the data for January and February 2022.

Read the data for January. How many columns are there?


*   16
*   17
*   18
*   19



In [1]:
import pandas as pd

taxi_df = pd.read_parquet('yellow_tripdata_2022-01.parquet')
print(taxi_df.shape)

(2463931, 19)


Note: dataframe shape returns a tuple showing the dimension of the dataframe. It shows the number of row and column: 2,463,931 rows and 19 columns

##Q2. Computing duration
Now let's compute the `duration` variable. It should contain the duration of a ride in minutes.

What's the standard deviation of the trips duration in January?

* 41.45
* 46.45
* 51.45
* 56.45

In [2]:
taxi_df.head(10)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1         1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2         2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3         2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4         2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   
5         1  2022-01-01 00:40:15   2022-01-01 01:09:48              1.0   
6         2  2022-01-01 00:20:50   2022-01-01 00:34:58              1.0   
7         2  2022-01-01 00:13:04   2022-01-01 00:22:45              1.0   
8         2  2022-01-01 00:30:02   2022-01-01 00:44:49              1.0   
9         2  2022-01-01 00:48:52   2022-01-01 00:53:28              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           3.80         1.0                  N           142           236   
1           2.10         1.0                  N           236            42   
2           0.97         1.0                  N           166           166   
3           1.09         1.0                  N           114            68   
4           4.30         1.0                  N            68           163   
5          10.30         1.0                  N           138           161   
6           5.07         1.0                  N           233            87   
7           2.02         1.0                  N           238           152   
8           2.71         1.0                  N           166           236   
9           0.78         1.0                  N           236           141   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             1         14.5    3.0      0.5        3.65          0.00   
1             1          8.0    0.5      0.5        4.00          0.00   
2             1          7.5    0.5      0.5        1.76          0.00   
3             2          8.0    0.5      0.5        0.00          0.00   
4             1         23.5    0.5      0.5        3.00          0.00   
5             1         33.0    3.0      0.5       13.00          6.55   
6             1         17.0    0.5      0.5        5.20          0.00   
7             2          9.0    0.5      0.5        0.00          0.00   
8             1         12.0    0.5      0.5        2.25          0.00   
9             2          5.0    0.5      0.5        0.00          0.00   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    0.3         21.95                   2.5          0.0  
1                    0.3         13.30                   0.0          0.0  
2                    0.3         10.56                   0.0          0.0  
3                    0.3         11.80                   2.5          0.0  
4                    0.3         30.30                   2.5          0.0  
5                    0.3         56.35                   2.5          0.0  
6                    0.3         26.00                   2.5          0.0  
7                    0.3         12.80                   2.5          0.0  
8                    0.3         18.05                   2.5          0.0  
9                    0.3          8.80                   2.5          0.0

In [4]:
taxi_df['duration'] = taxi_df.tpep_dropoff_datetime - taxi_df.tpep_pickup_datetime
taxi_df['duration'] = taxi_df.duration.dt.total_seconds() / 60
print("Standard Deviation for duration data: ", taxi_df['duration'].std())

Standard Deviation for duration data:  46.44530513776802


##Q3. Dropping outliers
Next, we need to check the distribution of the `duration` variable. There are some outliers. Let's remove them and keep only the records where the duration was between 1 and 60 minutes (inclusive).

What fraction of the records left after you dropped the outliers?

* 90%
* 92%
* 95%
* 98%

In [5]:
df = taxi_df[(taxi_df.duration >= 1) & (taxi_df.duration <= 60)]
print(df.shape)

(2421440, 20)


In [6]:
percentage = (df.shape[0])/taxi_df.shape[0]
print(f"{percentage:.2%}")

98.28%


In [7]:
df.head(5)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1         1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2         2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3         2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4         2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           3.80         1.0                  N           142           236   
1           2.10         1.0                  N           236            42   
2           0.97         1.0                  N           166           166   
3           1.09         1.0                  N           114            68   
4           4.30         1.0                  N            68           163   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             1         14.5    3.0      0.5        3.65           0.0   
1             1          8.0    0.5      0.5        4.00           0.0   
2             1          7.5    0.5      0.5        1.76           0.0   
3             2          8.0    0.5      0.5        0.00           0.0   
4             1         23.5    0.5      0.5        3.00           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  \
0                    0.3         21.95                   2.5          0.0   
1                    0.3         13.30                   0.0          0.0   
2                    0.3         10.56                   0.0          0.0   
3                    0.3         11.80                   2.5          0.0   
4                    0.3         30.30                   2.5          0.0   

    duration  
0  17.816667  
1   8.400000  
2   8.966667  
3  10.033333  
4  37.533333

##Q4. One-hot encoding
Let's apply one-hot encoding to the pickup and dropoff location IDs. We'll use only these two features for our model.

* Turn the dataframe into a list of dictionaries

* Fit a dictionary vectorizer

* Get a feature matrix from it

What's the dimensionality of this matrix (number of columns)?

* 2
* 155
* 345
* 515
* 715

In [8]:
print(df['PULocationID'].isnull().sum())
print(df['DOLocationID'].isnull().sum())

0
0


In [10]:
from sklearn.feature_extraction import DictVectorizer

cat = ['PULocationID', 'DOLocationID']
train_dicts = df[cat].astype('int').astype('str').to_dict(orient='records')
dv = DictVectorizer()
X_train = dv.fit_transform(train_dicts)
X_train.shape

(2421440, 515)

##Q5. Training a model
Now let's use the feature matrix from the previous step to train a model.

Train a plain linear regression model with default parameters

Calculate the RMSE of the model on the training data
What's the RMSE on train?

* 6.99
* 11.99
* 16.99
* 21.99

In [11]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

y_train = df.duration.values

lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred = lr.predict(X_train)

rmse  = mean_squared_error(y_train, y_pred, squared=False)

print(rmse)

6.986190742248472


##Q6. Evaluating the model
Now let's apply this model to the validation dataset (February 2022).

What's the RMSE on validation?

* 7.79
* 12.79
* 17.79
* 22.79

In [12]:
feb_taxi_df = pd.read_parquet('yellow_tripdata_2022-02.parquet')
feb_taxi_df['duration'] = feb_taxi_df.tpep_dropoff_datetime - feb_taxi_df.tpep_pickup_datetime
feb_taxi_df['duration'] = feb_taxi_df.duration.dt.total_seconds() / 60

feb_taxi_df.head(5)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2022-02-01 00:06:58   2022-02-01 00:19:24              1.0   
1         1  2022-02-01 00:38:22   2022-02-01 00:55:55              1.0   
2         1  2022-02-01 00:03:20   2022-02-01 00:26:59              1.0   
3         2  2022-02-01 00:08:00   2022-02-01 00:28:05              1.0   
4         2  2022-02-01 00:06:48   2022-02-01 00:33:07              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           5.40         1.0                  N           138           252   
1           6.40         1.0                  N           138            41   
2          12.50         1.0                  N           138           200   
3           9.88         1.0                  N           239           200   
4          12.16         1.0                  N           138           125   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             1         17.0   1.75      0.5        3.90          0.00   
1             2         21.0   1.75      0.5        0.00          6.55   
2             2         35.5   1.75      0.5        0.00          6.55   
3             2         28.0   0.50      0.5        0.00          3.00   
4             1         35.5   0.50      0.5        8.11          0.00   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  \
0                    0.3         23.45                   0.0         1.25   
1                    0.3         30.10                   0.0         1.25   
2                    0.3         44.60                   0.0         1.25   
3                    0.3         34.80                   2.5         0.00   
4                    0.3         48.66                   2.5         1.25   

    duration  
0  12.433333  
1  17.550000  
2  23.650000  
3  20.083333  
4  26.316667

In [13]:
feb_df = feb_taxi_df[(feb_taxi_df.duration >= 1) & (feb_taxi_df.duration <= 60)]

cat = ['PULocationID', 'DOLocationID']
feb_train_dicts = feb_df[cat].astype('int').astype('str').to_dict(orient='records')
feb_dv = DictVectorizer()
X_train_feb = feb_dv.fit_transform(feb_train_dicts)


X_train_feb.shape

(2918187, 514)

In [14]:
y_train_feb = feb_df.duration.values

lr = LinearRegression()
lr.fit(X_train_feb, y_train_feb)
y_pred_feb = lr.predict(X_train_feb)

feb_rmse  = mean_squared_error(y_train_feb, y_pred_feb, squared=False)

print(feb_rmse)

7.639501263098017
